### ISSUE TITLE
torch.nn.PairwiseDistance returns +Inf for moderately large p (overflow in fp32, no log-space rewrite)

### ISSUE THREAD

**Issue by gdrosa** [pytorch/pytorch#184036]
### 🐛 Describe the bug

`torch.nn.PairwiseDistance(p)` (forward delegates to `F.pairwise_distance`, which calls `torch.norm(x1 - x2 + eps, p, dim=-1)`) silently returns `+Inf` for every row when `p` is large enough that the inner `(|diff| + eps) ** p` saturates fp32 before the outer `** (1/p)` root can recover.

The threshold is much lower than absurd values. With unit-scale inputs (e.g. `torch.randn` differences with `max|diff| ~= 3`), overflow begins at `p ~= 76` and saturates **all** rows by `p ~= 95`. A hyperparameter sweep like `p in {1, 2, 5, 10, 50, 100, 1000}` silently gets `+Inf` for the upper half. The docstring at `torch/nn/modules/distance.py` describes the L-p norm `(Sum |x_i|^p)^(1/p)` whose limit as `p -> +inf` is the L-infinity norm `max|x_i|`; the implementation diverges from that limit.

`torch.linalg.vector_norm(diff, ord=p)` overflows in the same regime since it shares the underlying `torch.norm` code, so a fix benefits both APIs.

```python
import torch

torch.manual_seed(0)
x1 = torch.randn(4, 8, dtype=torch.float32)
x2 = torch.randn(4, 8, dtype=torch.float32)
for p in [50, 80, 100, 1000, 1e18]:
    pd = torch.nn.PairwiseDistance(p=p)
    print(f"p={p:>6}: {pd(x1, x2)}")
```

Observed:

```
p=    50: tensor([2.5577, 2.5867, 2.5688, 3.2271])
p=    80: tensor([2.5564,    inf, 2.5688, 3.2270])
p=   100: tensor([inf, inf, inf, inf])
p=  1000: tensor([inf, inf, inf, inf])
p= 1e+18: tensor([inf, inf, inf, inf])
```

Expected:

Either (a) raise `ValueError` (or warn) when `p * log(|diff|.max())` would exceed `log(dtype_max)`, or (b) compute the L-p norm in log space via `exp((1/p) * logsumexp(p * log(|diff| + eps)))`, which is finite for any `p > 0` and converges to `max|diff|` as `p -> +inf` (matching the docstring's mathematical limit).

### Versions

```
Collecting environment information...
PyTorch version: 2.12.0a0+git71a2161
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.11.15 | packaged by conda-forge | (main, Mar  5 2026, 16:45:40) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-111-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1660 SUPER
Nvidia driver version: 535.288.01
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  20
On-line CPU(s) list:                     0-19
Vendor ID:                               GenuineIntel
Model name:                              12th Gen Intel(R) Core(TM) i7-12700
CPU family:                              6
Model:                                   151
Thread(s) per core:                      2
Core(s) per socket:                      12
Socket(s):                               1
Stepping:                                2
CPU(s) scaling MHz:                      51%
CPU max MHz:                             4900.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4224.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities ibpb_exit_to_user
Virtualization:                          VT-x
L1d cache:                               512 KiB (12 instances)
L1i cache:                               512 KiB (12 instances)
L2 cache:                                12 MiB (9 instances)
L3 cache:                                25 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-19
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Mitigation; Clear Register File
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] optree==0.19.0
[pip3] torch==2.12.0a0+git71a2161
[pip3] triton==3.6.0
[conda] numpy                       2.4.4            pypi_0              pypi
```

cc @jianyuh @nikitaved @mruberry @walterddr @xwang233 @Lezcano

**Comment by puneetdixit200** [comment]
Hello maintainers, I would like to work on this numerical-stability issue. I can add regression tests for large p values in fp32, evaluate a stable computation strategy, and submit a focused fix. Please assign this to me.

**Comment by abcde2342** [comment]
I found that even with float64, the result is not good when p is large. Please take a look at my patch.

```
from __future__ import annotations

import math

from typing import Any

import mpmath as mp

import numpy as np
import torch

mp.mp.dps = 100

CASE_TITLE = 'torch.nn.PairwiseDistance returns +Inf for moderately large p (overflow in fp32, no log-space rewrite)'

CASE_FUNCTION = 'torch.nn.PairwiseDistance / vector_norm'

CASE_ID = 'pytorch/pytorch#184036'

FINITE_DIFF_FIXED_STEP = mp.mpf("1e-5")

def relerr(v, ref):
    is_complex = isinstance(v, complex) or isinstance(ref, (complex, mp.ctx_mp_python.mpc))
    v = mp.mpc(v) if is_complex else mp.mpf(v)
    ref = mp.mpc(ref) if is_complex else mp.mpf(ref)
    if abs(ref) == 0:
        return float(abs(v - ref))
    return float(abs((v - ref) / ref))

def orders(before, after):
    if not math.isfinite(before) or before <= 0:
        return 'inf' if after == 0 or math.isfinite(after) else 'nan'
    if after == 0:
        return 'inf'
    if not math.isfinite(after):
        return '-inf'
    return f'{math.log10(before / after):.2f}'

def pnorm_naive(vals, p):
    arr = torch.as_tensor(vals, dtype=torch.float64)
    x1 = arr.reshape(1, -1)
    x2 = torch.zeros_like(x1)
    out = torch.nn.PairwiseDistance(p=p)(x1, x2)
    return float(out)

def pnorm_taylor(vals, p, terms=6):
    arr = np.abs(np.asarray(vals, dtype=np.float64))
    
    scale = float(np.max(arr))
    if scale == 0:
        return 0.0
    ratios = arr / scale
    max_index = int(np.argmax(ratios))
    small = np.delete(ratios, max_index) ** float(p)
    s = float(np.sum(small))
    alpha = 1.0 / float(p)
    total = 1.0
    coeff = 1.0
    power = 1.0
    for n in range(1, terms + 1):
        coeff *= (alpha - (n - 1)) / n
        power *= s
        total += coeff * power
    return scale * total

def _mp_value(x):
    if isinstance(x, (complex, mp.ctx_mp_python.mpc)):
        return mp.mpc(x)
    return mp.mpf(str(float(x))) if isinstance(x, np.generic) else mp.mpf(str(x))

def _finite_difference_step(x):
    return FINITE_DIFF_FIXED_STEP

def _condition_from_vector_finite_difference(f, xs):
    xs = [_mp_value(x) for x in xs]
    fx = f(xs)
    values = []
    steps = []
    for i, xi in enumerate(xs):
        h = _finite_difference_step(xi)
        for step in (h, -h):
            try:
                xh = list(xs)
                xh[i] = xh[i] + step
                d = (f(xh) - fx) / step
                if mp.isfinite(abs(d)):
                    values.append(abs(d) if abs(fx) == 0 else abs(xi * d / fx))
                    steps.append(step)
                    break
            except Exception:
                continue
    if not values:
        return (math.nan, math.nan)
    return (float(max(values)), max(steps, key=lambda s: abs(s)))

def _mp_pnorm(xs, p):
    return mp.fsum([abs(_mp_value(v)) ** p for v in xs]) ** (mp.mpf(1) / p)

def _format_value(value: Any) -> str:
    try:
        if isinstance(value, (mp.ctx_mp_python.mpf, mp.ctx_mp_python.mpc)):
            return mp.nstr(value, 17)
        if isinstance(value, complex):
            return str(value)
        if isinstance(value, np.generic):
            return repr(value.item())
        return repr(value)
    except Exception:
        return str(value)

def case_pytorch_pytorch_184036() -> dict[str, Any]:
    case = 'pytorch/pytorch#184036'
    expressions = ('(sum(abs(x_i)**p))**(1/p)', 'mpmath p-norm', 'max(abs(x))*(1+s)**(1/p) binomial Taylor')
    vals = [2.5577, 2.5867, 2.5688, 3.2271]
    p = 100.0
    ref = mp.fsum([mp.mpf(str(abs(v))) ** p for v in vals]) ** (mp.mpf(1) / p)
    original_value = pnorm_naive(vals, p)
    taylor_value = pnorm_taylor(vals, p)
    before = relerr(original_value, ref)
    after = relerr(taylor_value, ref)
    condition = _condition_from_vector_finite_difference(lambda xs: _mp_pnorm(xs, mp.mpf('100')), vals)
    row = {'condition_number': condition[0], 'original_expression': expressions[0], 'oracle_expression': expressions[1], 'taylor_expression': expressions[2], 'original_value': _format_value(original_value), 'oracle_value': _format_value(ref), 'taylor_value': _format_value(taylor_value), 'original_relative_error': before, 'taylor_relative_error': after, 'taylor_orders_improved': orders(before, after)}
    return ([row], expressions[2])

def print_case_report() -> None:
    rows, patch_idea = case_pytorch_pytorch_184036()
    print(f"{CASE_ID}: {CASE_TITLE}")
    print(f"  function: {CASE_FUNCTION}")
    print(f"  patch: {patch_idea}")
    for row in rows:
        print("  evidence:")
        print(f"    condition_number={float(row['condition_number']):.6e}")
        print(f"    original_expression={row['original_expression']}")
        print(f"    oracle_expression={row['oracle_expression']}")
        print(f"    taylor_expression={row['taylor_expression']}")
        print(f"    original_value={row['original_value']}")
        print(f"    oracle_value={row['oracle_value']}")
        print(f"    taylor_value={row['taylor_value']}")
        print(f"    original_relative_error={float(row['original_relative_error']):.6e}")
        print(f"    taylor_relative_error={float(row['taylor_relative_error']):.6e}")
        print(f"    taylor_orders_improved={row['taylor_orders_improved']}")

if __name__ == "__main__":
    print_case_report()

```

**Comment by Dex947** [comment]
I looked into this and can reproduce it on current torch. With unit-scale fp32 inputs (`torch.randn` differences, `max|diff|` around 4-5), `F.pairwise_distance` returns `+Inf` for every row starting around `p = 76`, and all rows are saturated by `p ~= 95`. `torch.linalg.vector_norm(diff, ord=p)` overflows in the same regime, which makes sense since `pairwise_distance` is just a wrapper around `at::norm`.

The cause is the naive evaluation of `(Sum |d|^p)^(1/p)`: the inner `|d|^p` overflows fp32 long before the outer root could bring it back down. So the issue is really in the norm computation, not in `pairwise_distance` specifically.

The standard fix is to factor out the max before exponentiating:

```
m = max|d|
result = m * (Sum (|d| / m)^p)^(1/p)
```

I tested this against the naive version across a range of `p`:

| p | naive | max-factored | max\|diff\| |
|---|-------|--------------|-----------|
| 50 | 4.3125 | 4.3125 | 4.3013 |
| 76 | inf | 4.3066 | 4.3013 |
| 100 | inf | 4.3043 | 4.3013 |
| 1000 | inf | 4.3013 | 4.3013 |

It stays finite everywhere, matches the naive result where the naive one is still finite, and converges to `max|diff|` as `p -> inf`, which is the L-inf limit the docstring describes but the current code doesn't reach. The all-zero row needs a guard so `m = 0` doesn't produce `0/0`, but that's straightforward.

Before I take this further, two questions on scope:

1. Would you prefer this fixed narrowly in `pairwise_distance`, or properly in `at::norm` / `linalg_vector_norm` since they share the overflow? My concern with the broad fix is the common `p = 2` path: the max-factoring adds a reduction and a division, so I'd want to gate it to avoid any regression on the hot path, and back that with a benchmark.

2. @abcde2342 posted a patch in a comment above a couple of weeks ago. Are you still working on this? Happy to take it forward with a fix and regression tests if not, otherwise glad to help on yours.
